2019/02/14.md

```diff
@@ -0,0 +1,199 @@
+# `dhis2/ui` feedback meeting
+
+## Attending
+
+- austin
+- hendrik
+- jennifer
+- viktor
+
+## Re-export files
+
+- Webpack 4 implements strict es modules for mjs files
+- Unclear if that's an actual specification
+
+**Investigation results**: 
+
+The eco-system surrounding JavaScript modules is split but there is work
+to bring the two biggest ones: ECMAScript modules and CommonJS, closer
+together[1].
+
+Webpack 4 has implemented something called strict-ESM, which basically
+enforces[2]:
+
+>Dynamic modules (non-esm, i. e. CommonJs) can only imported via default
+>import, everything else (including namespace import) emit errors
+
+There is additional information straight from the webpack team in [3] as
+well: 
+
+>webpack supports the `__esModule` flag, with upgrades a CJS module to a ESM.
+
+It does not upgrade a `__esModule` module to a strict-ESM module
+however:
+
+>In strict-ESM we don’t allow picking properties via `import` and only
+>allow the default export for non-ESM.
+
+This breaks in UI.
+
+- [1] https://github.com/nodejs/modules
+- [2] https://medium.com/@zwegrzyniak/webpack-4-1-and-es-modules-esm-dd0bd7dca4da
+- [3] https://medium.com/webpack/webpack-4-import-and-commonjs-d619d626b655
+
+**Conclusion**: Now that the cause is clear, we need to decide on what
+direction to take with the UI-library. All the approaches has drawbacks.
+One thing is clear, if we are going to use ESM compiled to CJS as it
+stands today, we do not use `.mjs` as the file extension.
+
+## Fonts
+
+- The application font will be used by the application as well as the
+  components
+- We don't want to run the risk of bundling it twice or defining it
+  twice
+
+**Conclusion**:
+
+- Use MDIDX for icons for v1.0
+- As an optimization step we should switch to SVGs, this triggers
+  changes in the `Icon` component, but it should be a non-breaking
+  change.
+- The host application will pull in application level fonts (e.g.
+  Roboto), but would be nice to throw some sort of error if it doesn't.
+- The UI components expect that the font is loaded by that application
+  and just defines the `font-family` it needs
+
+## Feedback
+
+Input from the field.
+
+### Down-side of isolated styles
+
+- Difficult to re-use primitives of styles
+
+**Note**:
+
+- Some of the primitives, for example the colors, are exposed from the
+  UI library as CSS, which can be imported and re-used.
+
+**Conclusion**: This is something to watch out for. If this setup adds
+duplication of colors and standard styles throughout all applications we
+need to change it in `dhis2/ui`.
+
+### Down-side of locked down apis
+
+- No paddings/margins atm, so everything has been wrapped in another
+  components to allow for whitespace.
+
+**Conclusion**:
+
+- There is a design for a grid system which allows a whitelisted set of
```

Comment:
Tangent - how does this work when you might want different x and y padding or margin?  This approach sounds overly-complex to me, and is just asking for a million wrapper divs.

Reply:
Good question, I'll ask Joe what he had in mind for his grid.

Reply:
Yeah, we need something better than defining the padding/margin all-around on a single prop. Maybe open up for a `className` prop instead on components.

Reply:
I think the `className` prop would be a pragmatic choice. I do like the grid idea in principle, but I think in practice it will backfire:
- On the one hand we'll end up with apps wit a lot of wrapper divs because the grid system doesn't cover the scenario in question, and the developer just wants to proceed building the app.
- On the other hand we'll end up continuously expanding the whitelist while more justifiable style customisations emerge. It would eat up a lot of our time. As an illustration of what I mean, take the release notes of React 16. [Maintaining a whitelist](https://reactjs.org/blog/2017/09/08/dom-attributes-in-react-16.html#why-are-we-changing-this) didn't work for them either.

Allowing a `className` prop gives great power. And with great power comes great responsibility. I think most devs will take that responsibility.

Reply:
Using the PostCSS approach makes the `className` prop option difficult because it is unpredictable. The scope of the CSS which is generated for the component is so strict that any overrides are difficult.

I'm prototyping using `styled-jsx` instead. It allows for the `className` prop option to be used too.